### PR TITLE
Remove aggregation's postCollect phase

### PR DESCRIPTION
**Pull request by nik9000**
After #63811 it became clear to me that `postCollect` is kind of
dangerous and not all that useful. So this removes it.

The trouble with `postCollect` is that it all happened right after we
finished calling `collect` on the `LeafBucketCollectors` but before we
built the aggregation results. But in #63811 we found out that we can't
call `postCollect` on the children of `parent` or `child` aggregators
until we know which *which* aggregation results we're building.

So this removes `postCollect` and moves all of the things we did at
post-collect phase into `buildAggregations` or into hooks called in
those methods.

### CODE REVIEW

File: modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ParentJoinAggregator.java [hunk 1]
```diff
@@ -113,12 +113,7 @@ public void collect(int docId, long owningBucketOrd) throws IOException {
     }
 
     @Override
-    public void postCollection() throws IOException {
-        // Delaying until beforeBuildingBuckets
-    }
-
-    @Override
-    protected void beforeBuildingBuckets(long[] ordsToCollect) throws IOException {
+    protected void prepareSubAggs(long[] bucketOrdsToCollect) throws IOException {
```
Comment (nik9000) [inline modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ParentJoinAggregator.java:116]: I renamed this method to make it a bit more clear what it is for.

File: server/src/main/java/org/elasticsearch/search/aggregations/Aggregator.java [hunk 2]
```diff
@@ -142,12 +142,12 @@ public BucketComparator bucketComparator(String key, SortOrder order) {
 
     /**
      * Build the results of this aggregation.
-     * @param owningBucketOrds the ordinals of the buckets that we want to
+     * @param ordsToCollect the ordinals of the buckets that we want to
      *        collect from this aggregation
      * @return the results for each ordinal, in the same order as the array
      *         of ordinals
      */
-    public abstract InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/Aggregator.java:150]: I *think* this is a more clear variable name. I made this change in a dead end when I was working on this change locally but I kind of like it.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java [hunk 3]
```diff
@@ -136,20 +135,12 @@ public void preCollection() throws IOException {
...
-
     /**
      * Replay the wrapped collector, but only on a selection of buckets.
      */
     @Override
     public void prepareSelectedBuckets(long... selectedBuckets) throws IOException {
-        if (finished == false) {
-            throw new IllegalStateException("Cannot replay yet, collection is not finished: postCollect() has not been called");
-        }
+        finishLeaf();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java:143]: I'm pretty happy about how this turned out! No more `finished` to check on. It's just all built in now.

File: server/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java [hunk 4]
```diff
@@ -124,17 +124,6 @@ public void preCollection() throws IOException {
         profiler.pollLastElement();
     }
 
-    @Override
-    public void postCollection() throws IOException {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/profile/aggregation/ProfilingAggregator.java:128]: This is kind of an unfortunate loss here. Now we'll end up having `post_collection` in the `build_aggregation` debug phase. I'm wondering if I can have aggregations that do "interesting" things opt in to make their own timers or something.

**Comment by nik9000** [comment]
@cjcenizal is this something that you need to look at as well? It removes one of the entries in the aggs profile list because that line was "scary".

**Comment by cjcenizal** [comment]
@nik9000 Thanks for the ping! Which API does this affect, and what is the change? Is it in the request or the response? Could you provide an example before and after?

**Comment by nik9000** [comment]
> @nik9000 Thanks for the ping! Which API does this affect, and what is the change? Is it in the request or the response? Could you provide an example before and after?

Just the response. Have a look at the asciidoc file this PR changes.

**Review by cjcenizal — APPROVED** [review]
Doesn't look like ES UI depends on the `post_collection` or `post_collection_count` fields which are removed from the `profile` response in this PR.

**Comment by nik9000** [comment]
> Doesn't look like ES UI depends on the `post_collection` or `post_collection_count` fields which are removed from the `profile` response in this PR.

@cjcenizal would the UI be ok if I added fields to that list? Renamed them? I've been thinking about having a little more flexibility there.

**Comment by nik9000** [comment]
run elasticsearch-ci/bwc

**Review by not-napoleon — APPROVED** [review]
LGTM

**Comment by cjcenizal** [comment]
@nik9000 

> would the UI be ok if I added fields to that list? Renamed them? I've been thinking about having a little more flexibility there.

From what I can see, the code builds the UI recursively and dynamically, following the structure of the Profile API response but agnostic to its contents. So you should be able to add, remove, and rename fields freely and I expect the UI to faithfully surface them.

image

**Comment by nik9000** [comment]
> From what I can see, the code builds the UI recursively and dynamically, following the structure of the Profile API response but agnostic to its contents. So you should be able to add, remove, and rename fields freely and I expect the UI to faithfully surface them.

:metal: 

That means if an aggs has an "unusual" phase it can add it rather than tucking it into the other phases. That seems useful!

**Pull request merged**